### ISSUE TITLE
Tinami Uploading Support

### ISSUE THREAD

**Issue by BrokenEagle** [danbooru/danbooru#2818]
There currently exists no support for Tinami, making it necessary to download images locally, and then uploading from disk and filling in the source manually.

- [ ] Step 1: Add support for Tinami image URLs.

When using the bookmarklet on a Tinami image URL, it apparently seems to work at first, as it does pull the image up on the preview.  However, when attempting to submit that upload, it always fails with the HTTP 403 status code.  This can be seen at http://danbooru.donmai.us/uploads/1277085. ([Backed up on Google Drive](https://drive.google.com/open?id=0B39jDgCtzVHudVVEV2FsZko0NFE))

`An error occurred: error: Downloads::File::Error - HTTP error code: 403 Forbidden`

After some testing, I discovered that the "Referer" field needs to be set to the following in order to download  images successfully from Tinami.

`tinamiheader = {
    'Referer':'http://www.tinami.com/'}`

- [ ] Step 2: Add support for Tinami post URLs Part 1 (For one or more images with no popup)

Images without a popup will not have 'form method="post"' as part of the HTML structure that surrounds Tinami URL image links. Those image links are of the following format as demonstrated through REGEX (single quote delimited).

`'http://img.tinami.com/illust\d?/img/[^/]+/[^"]+'`

**Note:** Some checking does need to occur to verify that there are no popup links (with the POST method) to larger full-size image, to avoid uploading the sample size instead.

- [ ] Step 3: Add support for Tinami post URLs Part 2 (With one or more images that have fullsizes available via popup link)

If there is a popup link as indicated by the presence of  'form method="post"' in the HTML as part of a displayed image, then a multistage process needs to be followed to get the fullsize image link.

1. Do a GET request to the post page with authentication

Post URLs take the format of http://www.tinami.com/view/<POSTID>.  The only requirement at this point for the GET request is to set the authentication cookie, "rem2", to your particular accounts authentication token.

`tinamiheader={
    'Cookie': "rem2=AAAA7d5aBBBB409dCCCC595122cc3ccc"}`

**Note:** Token above is not real and for demonstration purposes only

The response header will have the following information for the field "Set-Cookie".

```
{"Set-Cookie" : "Tinami2SESSID=24be021ef670eb1adde26f43ef60a0d688d26e5e147abc14b075de3c1cd2b15e; path=/, Tinami2SESSID=74528d719554e1cd2a0ad189bb31ed92; path=/, authchallenge=28e8bc6eeea8118364c7bad302cad7fc; path=/"}
```

The only important cookie here is the second one, or the 32-character hexadecimal "Tinami2SESSID" cookie, however not the 64-character hexadecimal "Tinami2SESSID" that precedes it.

From the HTML, the primary value of importance called "ethna_csrf" (also a 32-character hexadecimal), which can be found with the following REGEX (delimited by single quotes).

`'name="ethna_csrf" value="([^"]+)"'`

**Note:** Some post images have an extra input value called "no", with a value usually starting at 1, and incrementing from there depending upon how many popup images there are.

2. Do a POST request to the POST page with authentication

For the header, the following cookies and referral portions need to be filled out.
```
tinamiheader = {
    'Referer': 'http://www.tinami.com/'
    'Cookie': 'rem2=AAAA7d5aBBBB409dCCCC595122cc3ccc; Tinami2SESSID=74528d719554e1cd2a0ad189bb31ed92'}
```
For the data, the following data items need to be set.
```
tinamidata = {
	'action_view_original': 'true', 
	'ethna_csrf': 'VALIDATIONTOKEN', 
	'cont_id': 'POSTID'}
```
**Note:** If the image had a data value called "no", it needs to be added to the data section above.  However, even if there is more than one "no" value, the "ethna_csrf" value will remain the same for all images from that post page.

Where the VALIDATIONTOKEN is the value captured for "ethna_csrf" and POSTID is the post id as part of the URL. An example data body URL encoded looks like the following
`action_view_original=true&ethna_csrf=d63a47de529e85027b4ff5932e639427&cont_id=787128`

3. Get the full-size image URL from the post response.

If the above actions are successful, then the HTML response should be short and look like the following.
```
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">

<title>TINAMI - ［イラスト］ダメージド響</title>
</head>
<div style="text-align:center;margin:1em auto;">
<a href="#" onClick="window.close(); return false;"><img src="http://img.tinami.com/illust2/img/818/55952dd89a424.jpg" style="border:none;"></a>
</div>
</body>
</html>
```
However, some verification needs to be done that the POST was successful as it can fail in two ways.  The first will just redirect as a GET request to the post page. The second will redirect to a general failure page.

**Comment by r888888888** [comment]
I don't think anyone uses Tinami anymore.

**Comment by evazion** [comment]
Fixed in 01b683798eeab2abceaf9a36413376afd1e10007. As far as I can tell authentication isn't required.
